content-services/content-sources-backend
# Fixes 1939: pulp domain integration

## Summary

* Updates to use our own docker-compose for pulp
* Create a domain via a go routine at api calls to update/create if snapshotting is to be performed
* Split the pulp client into 2 interfaces, a global and a non-global.  Global doesn't require a domain.  Underlying struct is the same.

NOTE:  Before merging, we need to set  `DOMAIN_ENABLED: "true"` in the pulp deployment for ephemeral, within our app-inteface deploy.yaml:
```
  - namespace:
      $ref: /services/insights/ephemeral/namespaces/ephemeral-base.yml
    disable: true  # do not create an app-sre deploy job for ephemeral namespace
    ref: internal
    parameters:
      IMAGE_TAG: latest
```
just add below IMAGE_TAG:
```
DOMAIN_ENABLED: "true"
```

## Testing steps

 * make sure to be on main
 * make compose-clean
 * checkout this pr
 * make compose-up
 * create a repo:
 ```
curl -X POST --location "http://localhost:8000/api/content-sources/v1.0/repositories/" \
    -H "Content-Type: application/json" \
    -H "x-rh-identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiVXNlciIsInVzZXIiOnsidXNlcm5hbWUiOiJqZG9lIn0sImludGVybmFsIjp7Im9yZ19pZCI6IjEyMyJ9fX0K" \
    -d "{\"name\":\"pulp 3.14\",\"url\":\"http://yum.theforeman.org/pulpcore/3.14/el8/x86_64/\",\"snapshot\": true}"
```
Use a bulk create with a different org:
```
curl -X POST --location "http://localhost:8000/api/content-sources/v1.0/repositories/bulk_create/"    -H "Content-Type: application/json"    -H "x-rh-identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiVXNlciIsInVzZXIiOnsidXNlcm5hbWUiOiJhc2RmYXNkZiJ9LCJhY2NvdW50X251bWJlciI6ImFhc2RmIiwiaW50ZXJuYWwiOnsib3JnX2lkIjoiYWFzZGYifX19Cg=="    -d "[{\"name\":\"pulp 3.14\",\"url\":\"http://yum.theforeman.org/pulpcore/3.14/el8/x86_64/\",\"snapshot\": true}, {\"name\":\"pulp 3.16\",\"url\":\"http://yum.theforeman.org/pulpcore/3.16/el8/x86_64/\",\"snapshot\": true}]"
```

List the snapshots for the repo and see that the new 'RepositoryPath' shows a domain (we still store the distribution path in the db, but thats not relevant to the user)

Also, do a bunch of bulk creates on a new org, verify that there are no errors.

### Ephemeral testing:
1.  Create a bonfire config:
```
appsFile:
  host: gitlab
  repo: insights-platform/cicd-common
  path: bonfire_configs/ephemeral_apps.yaml
apps:
 - name: content-sources
   components:
     - name: pulp
       host: github
       repo: pulp/pulp-clowder-deployments
       path: deploy/clowdapp.yaml
       parameters:
         DOMAIN_ENABLED: 'true'
     - name: content-sources-backend
       host: local
       repo: /home/jlsherri/git/content-sources-backend
       path: deployments/deployment.yaml
       parameters:
         IMAGE_TAG: pr-331-latest
         FEATURES_ADMIN_TASKS_ENABLED: true
         FEATURES_SNAPSHOTS_ENABLED: true
     - name: content-sources-frontend
       host: github
       repo: content-services/content-sources-frontend

```

reserve a namespace and deploy:
```bonfire deploy content-sources --frontends=true```

Once its up do a `namespace describe` to get the info and curl to create a repo (replacing with the correct info):
```
 curl -v -u jdoe:OK4dkBMZ7CKhZ8lH   -X POST https://env-ephemeral-bvyuo3-d7v745fk.apps.c-rh-c-eph.8p0c.p1.openshiftapps.com/api/content-sources/v1.0/repositories/  -d '{"name":"pulp 3.18","url":"http://yum.theforeman.org/pulpcore/3.18/el8/x86_64/","snapshot": true}' -H 'Content-Type: application/json'
```
Monitor the admin tasks, and verify that the task is created.  You can curl the snapshots api to confirm the path:

```
curl -v -u jdoe:OSRFgOKvL5ZLdQbO https://env-ephemeral-7sp1sd-1o8n8w1w.apps.c-rh-c-eph.8p0c.p1.openshiftapps.com/api/content-sources/v1.0/repositories/21202889-e9a2-4edf-87fa-b673c5f0339c/snapshots/ | jq
```

for example, my output shows a repository_path of: 
```
    "repository_path_path": "bf7f7fae/21202889-e9a2-4edf-87fa-b673c5f0339c/b959ce1f-7b6c-467c-857c-6afa3808abd6",
```
that first char set `bf7f7fae` is the domain name.

you might also try bulk creation here

### testing with local minion:
 * Login to minio at http://localhost:9001
 * create a bucket, named test
 * create a user (under identity) with test/password.  Give them the read/write role
 * add this to your config.yaml:
 ```  
pulp:
    server: http://localhost:8080
    username: admin
    password: password
    storage_type: object #object or local
    custom_repo_objects:
      url: http://minio:9000
      accessKey: test
      secretKey: password
      name: test
      region: rdu
```
* Create a repo in a new organization id (domains aren't currently updated).
